### PR TITLE
fix: [PL-29986]: added check for JWT verification before masking

### PR DESCRIPTION
**Pull request by tiwaritejas**
## Describe your changes
In order to avoid overmatching for JWT regex during masking, a check to decode JWT has been introduced. Only if the the regex is a valid JWT will it get masked.
## Checklist
- [x] I've documented the changes in the PR description.
- [x] I've tested this change either in PR or local environment.
- [x] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>

You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
    - CodeFormat: `trigger codeformat`
    - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- PMD: `trigger pmd`
- Feature Name Check: `trigger featurenamecheck`
- TI-bootstrap: `trigger ti0`
- TI-bootstrap1: `trigger ti1`
- TI-bootstrap2: `trigger ti2`
- TI-bootstrap3: `trigger ti3`
- TI-bootstrap4: `trigger ti4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- Trigger all Checks: `trigger smartchecks`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/40549)

### CODE REVIEW

**Comment by mohitg0795** [comment]
*[`930-delegate-tasks/src/main/java/software/wings/delegatetasks/LogSanitizer.java` line 78 at r1](https://reviewable.io/reviews/harness/harness-core/40549#-NHa5laV9DDEeC0Ira_V:-NHa5laV9DDEeC0Ira_W:b-2ff0ka) ([raw file](https://github.com/harness/harness-core/blob/69bda0209b1b10ad5b800eb632c853a10a4269b7/930-delegate-tasks/src/main/java/software/wings/delegatetasks/LogSanitizer.java#L78)):*
> ```Java
> 
>   private String sanitizeJWT(String message) {
>     Pattern pattern = Pattern.compile(JWT_REGEX);
> ```

how about making this as static variable in the class so that we don't need to instantiate it again and again?

**Comment by mohitg0795** [comment]
*[`930-delegate-tasks/src/main/java/software/wings/delegatetasks/LogSanitizer.java` line 88 at r1](https://reviewable.io/reviews/harness/harness-core/40549#-NHa5rMG9NydiSWeZZds:-NHa5rMG9NydiSWeZZdt:bovsp72) ([raw file](https://github.com/harness/harness-core/blob/69bda0209b1b10ad5b800eb632c853a10a4269b7/930-delegate-tasks/src/main/java/software/wings/delegatetasks/LogSanitizer.java#L88)):*
> ```Java
>         JWT.decode(regexMatch);
>         message = message.replaceAll(regexMatch, SECRET_MASK);
>       } catch (JWTDecodeException ex) {
> ```

log exception

**Comment by mohitg0795** [comment]
*[`930-delegate-tasks/src/main/java/software/wings/delegatetasks/LogSanitizer.java` line 79 at r1](https://reviewable.io/reviews/harness/harness-core/40549#-NHa65ma389gWnQIKgp7:-NHa65ma389gWnQIKgp8:b-uuteag) ([raw file](https://github.com/harness/harness-core/blob/69bda0209b1b10ad5b800eb632c853a10a4269b7/930-delegate-tasks/src/main/java/software/wings/delegatetasks/LogSanitizer.java#L79)):*
> ```Java
>   private String sanitizeJWT(String message) {
>     Pattern pattern = Pattern.compile(JWT_REGEX);
>     String finalMessage = message;
> ```

whats the need for finalMessage variable? I guess this can be avoided, please CMIIW. Also, we can use better name as it indicates as if it will be the final response message while we are actually returning "message" as final response.

**Comment by mohitg0795** [comment]
*[`950-log-client/src/main/java/io/harness/logstreaming/LogStreamingSanitizer.java` line 50 at r1](https://reviewable.io/reviews/harness/harness-core/40549#-NHa6N2hE_1MM0pGhlPF:-NHa6N2hE_1MM0pGhlPG:bj86ql4) ([raw file](https://github.com/harness/harness-core/blob/69bda0209b1b10ad5b800eb632c853a10a4269b7/950-log-client/src/main/java/io/harness/logstreaming/LogStreamingSanitizer.java#L50)):*
> ```Java
>   }
> 
>   private String sanitizeJWT(String message) {
> ```

This is exact duplicate code. If any later changes are required, we would need to always do at both places. Such code should come from common libs only. Please place it inside 980-commons or some common lib which can be available to all services (if its multi-service use-case) for use.

**Review by nishantsaini20 — CHANGES_REQUESTED** [review]
> Reviewed 2 of 4 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/40549)* status: 4 file(s) to review. 1 change requests. 2. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`930-delegate-tasks/src/main/java/software/wings/delegatetasks/LogSanitizer.java` line 88 at r1](https://reviewable.io/reviews/harness/harness-core/40549#-NHa5rMG9NydiSWeZZds:-NHcQRqYA-XqNP1Mm6SQ:b-wo7in9) ([raw file](https://github.com/harness/harness-core/blob/69bda0209b1b10ad5b800eb632c853a10a4269b7/930-delegate-tasks/src/main/java/software/wings/delegatetasks/LogSanitizer.java#L88)):*
<details><summary><i>Previously, mohitg0795 (Mohit Garg) wrote…</i></summary><blockquote>

log exception
</blockquote></details>

We don't need to log this because if the matched string is not a JWT token it will fails and in that case we need to skip masking. This can drastically increase unnecessary logs.
___
*[`930-delegate-tasks/src/main/java/software/wings/delegatetasks/LogSanitizer.java` line 90 at r1](https://reviewable.io/reviews/harness/harness-core/40549#-NHcQoz01rtKyn1MHKi1:-NHcQoz01rtKyn1MHKi2:b-pp0rpa) ([raw file](https://github.com/harness/harness-core/blob/69bda0209b1b10ad5b800eb632c853a10a4269b7/930-delegate-tasks/src/main/java/software/wings/delegatetasks/LogSanitizer.java#L90)):*
> ```Java
>       } catch (JWTDecodeException ex) {
>         continue;
>       }
> ```

Add a generic Exception  catch block and log the error. Logging should not fail because of uncaught error caused by JWT.decode()
___
*[`950-log-client/src/main/java/io/harness/logstreaming/LogStreamingSanitizer.java` line 50 at r1](https://reviewable.io/reviews/harness/harness-core/40549#-NHa6N2hE_1MM0pGhlPF:-NHcRDRd6dVoO8w_xErg:b-bzgaf9) ([raw file](https://github.com/harness/harness-core/blob/69bda0209b1b10ad5b800eb632c853a10a4269b7/950-log-client/src/main/java/io/harness/logstreaming/LogStreamingSanitizer.java#L50)):*
<details><summary><i>Previously, mohitg0795 (Mohit Garg) wrote…</i></summary><blockquote>

This is exact duplicate code. If any later changes are required, we would need to always do at both places. Such code should come from common libs only. Please place it inside 980-commons or some common lib which can be available to all services (if its multi-service use-case) for use.
</blockquote></details>

Agree. This should be done

**Review by tiwaritejas — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/40549)* status: 6 file(s) to review. 1 change requests. 5. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`930-delegate-tasks/src/main/java/software/wings/delegatetasks/LogSanitizer.java` line 78 at r1](https://reviewable.io/reviews/harness/harness-core/40549#-NHa5laV9DDEeC0Ira_V:-NHd8bXYBnNTVqtNJk9-:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/69bda0209b1b10ad5b800eb632c853a10a4269b7/930-delegate-tasks/src/main/java/software/wings/delegatetasks/LogSanitizer.java#L78)):*
<details><summary><i>Previously, mohitg0795 (Mohit Garg) wrote…</i></summary><blockquote>

how about making this as static variable in the class so that we don't need to instantiate it again and again?
</blockquote></details>

Done.
___
*[`930-delegate-tasks/src/main/java/software/wings/delegatetasks/LogSanitizer.java` line 79 at r1](https://reviewable.io/reviews/harness/harness-core/40549#-NHa65ma389gWnQIKgp7:-NHd8cgiDyh8gzMVRW2m:bs2y0jo) ([raw file](https://github.com/harness/harness-core/blob/69bda0209b1b10ad5b800eb632c853a10a4269b7/930-delegate-tasks/src/main/java/software/wings/delegatetasks/LogSanitizer.java#L79)):*
<details><summary><i>Previously, mohitg0795 (Mohit Garg) wrote…</i></summary><blockquote>

whats the need for finalMessage variable? I guess this can be avoided, please CMIIW. Also, we can use better name as it indicates as if it will be the final response message while we are actually returning "message" as final response.
</blockquote></details>

It is required since we are using it in a lambda expression- https://stackoverflow.com/questions/34865383/variable-used-in-lambda-expression-should-be-final-or-effectively-final
___
*[`930-delegate-tasks/src/main/java/software/wings/delegatetasks/LogSanitizer.java` line 88 at r1](https://reviewable.io/reviews/harness/harness-core/40549#-NHa5rMG9NydiSWeZZds:-NHd8oBv9Bmz08zxUTQO:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/69bda0209b1b10ad5b800eb632c853a10a4269b7/930-delegate-tasks/src/main/java/software/wings/delegatetasks/LogSanitizer.java#L88)):*
<details><summary><i>Previously, nishantsaini20 (Nishant Saini) wrote…</i></summary><blockquote>

We don't need to log this because if the matched string is not a JWT token it will fails and in that case we need to skip masking. This can drastically increase unnecessary logs.
</blockquote></details>

Done.
___
*[`930-delegate-tasks/src/main/java/software/wings/delegatetasks/LogSanitizer.java` line 90 at r1](https://reviewable.io/reviews/harness/harness-core/40549#-NHcQoz01rtKyn1MHKi1:-NHd8pvw-0rFkhA5BK_E:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/69bda0209b1b10ad5b800eb632c853a10a4269b7/930-delegate-tasks/src/main/java/software/wings/delegatetasks/LogSanitizer.java#L90)):*
<details><summary><i>Previously, nishantsaini20 (Nishant Saini) wrote…</i></summary><blockquote>

Add a generic Exception  catch block and log the error. Logging should not fail because of uncaught error caused by JWT.decode()
</blockquote></details>

Done.
___
*[`950-log-client/src/main/java/io/harness/logstreaming/LogStreamingSanitizer.java` line 50 at r1](https://reviewable.io/reviews/harness/harness-core/40549#-NHa6N2hE_1MM0pGhlPF:-NHd8sTOEIGpUOvYPNJ0:b-9ur3te) ([raw file](https://github.com/harness/harness-core/blob/69bda0209b1b10ad5b800eb632c853a10a4269b7/950-log-client/src/main/java/io/harness/logstreaming/LogStreamingSanitizer.java#L50)):*
<details><summary><i>Previously, nishantsaini20 (Nishant Saini) wrote…</i></summary><blockquote>

Agree. This should be done
</blockquote></details>

Refactored into a helper method

**Review by nishantsaini20 — CHANGES_REQUESTED** [review]
> Reviewed 3 of 4 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/40549)* status: 6 file(s) to review. 1 change requests. 4. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`980-commons/src/main/java/io/harness/logging/LogSanitizerHelper.java` line 15 at r2](https://reviewable.io/reviews/harness/harness-core/40549#-NHdADtxCvWVha8EjXNF:-NHdADtxCvWVha8EjXNG:bctwdwc) ([raw file](https://github.com/harness/harness-core/blob/0f0e28539d09586f9aa318f9b7d9721c40166fa9/980-commons/src/main/java/io/harness/logging/LogSanitizerHelper.java#L15)):*
> ```Java
> public class LogSanitizerHelper {
>   public static String sanitizeJWT(String message) {
>     final String JWT_REGEX = "[\\w-]*\\.[\\w-]*\\.[\\w-]*";
> ```

This can be class constant
___
*[`980-commons/src/main/java/io/harness/logging/LogSanitizerHelper.java` line 16 at r2](https://reviewable.io/reviews/harness/harness-core/40549#-NHdAEbu6TZ1f6lP5UhQ:-NHdA-64AHOTj_XQUJB2:b-rwvll0) ([raw file](https://github.com/harness/harness-core/blob/0f0e28539d09586f9aa318f9b7d9721c40166fa9/980-commons/src/main/java/io/harness/logging/LogSanitizerHelper.java#L16)):*
> ```Java
>   public static String sanitizeJWT(String message) {
>     final String JWT_REGEX = "[\\w-]*\\.[\\w-]*\\.[\\w-]*";
>     Pattern pattern = Pattern.compile(JWT_REGEX);
> ```

This can be class static final
___
*[`980-commons/src/main/java/io/harness/logging/LogSanitizerHelper.java` line 27 at r2](https://reviewable.io/reviews/harness/harness-core/40549#-NHdAVBaEGHAme9ivJG7:-NHdAVBaEGHAme9ivJG8:b4myonj) ([raw file](https://github.com/harness/harness-core/blob/0f0e28539d09586f9aa318f9b7d9721c40166fa9/980-commons/src/main/java/io/harness/logging/LogSanitizerHelper.java#L27)):*
> ```Java
>         message = message.replaceAll(regexMatch, SECRET_MASK);
>       } catch (Exception ex) {
>         if (ex instanceof JWTDecodeException) {
> ```

can be two catch blocks as:



_Code snippet:_
```Java
catch (JWTDecodeException ex){
  continue;
} catch (Exception ex) {
  log.error("Error while trying to decode JWT", ex);
}
```

**Comment by tiwaritejas** [comment]
trigger ss

**Review by tiwaritejas — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/40549)* status: 6 file(s) to review. 1 change requests. 7. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`980-commons/src/main/java/io/harness/logging/LogSanitizerHelper.java` line 15 at r2](https://reviewable.io/reviews/harness/harness-core/40549#-NHdADtxCvWVha8EjXNF:-NHdKmRJBzITLtlSORDK:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/0f0e28539d09586f9aa318f9b7d9721c40166fa9/980-commons/src/main/java/io/harness/logging/LogSanitizerHelper.java#L15)):*
<details><summary><i>Previously, nishantsaini20 (Nishant Saini) wrote…</i></summary><blockquote>

This can be class constant
</blockquote></details>

Done.
___
*[`980-commons/src/main/java/io/harness/logging/LogSanitizerHelper.java` line 16 at r2](https://reviewable.io/reviews/harness/harness-core/40549#-NHdAEbu6TZ1f6lP5UhQ:-NHdKmfu210HJxbDaIAi:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/0f0e28539d09586f9aa318f9b7d9721c40166fa9/980-commons/src/main/java/io/harness/logging/LogSanitizerHelper.java#L16)):*
<details><summary><i>Previously, nishantsaini20 (Nishant Saini) wrote…</i></summary><blockquote>

This can be class static final
</blockquote></details>

Done.
___
*[`980-commons/src/main/java/io/harness/logging/LogSanitizerHelper.java` line 27 at r2](https://reviewable.io/reviews/harness/harness-core/40549#-NHdAVBaEGHAme9ivJG7:-NHdKnBzCBnIOnAGuQaq:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/0f0e28539d09586f9aa318f9b7d9721c40166fa9/980-commons/src/main/java/io/harness/logging/LogSanitizerHelper.java#L27)):*
<details><summary><i>Previously, nishantsaini20 (Nishant Saini) wrote…</i></summary><blockquote>

can be two catch blocks as:
</blockquote></details>

Done.

**Comment by tiwaritejas** [comment]
trigger codebasehash

**Review by nishantsaini20 — APPROVED** [review]
> Reviewed 1 of 4 files at r2, 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/40549)* status: 6 file(s) to review. 4. unresolved discussions. 1 of 2 approvals obtained. 

___
*[`980-commons/src/main/java/io/harness/logging/LogSanitizerHelper.java` line 26 at r5](https://reviewable.io/reviews/harness/harness-core/40549#-NHdg8VqB76Tx96RILAo:-NHdg8VrE5dqZuNgy7D-:bdojev8) ([raw file](https://github.com/harness/harness-core/blob/46263a816c6f5d6b270101e4181f30f1c720e210/980-commons/src/main/java/io/harness/logging/LogSanitizerHelper.java#L26)):*
> ```Java
>         JWT.decode(regexMatch);
>         message = message.replaceAll(regexMatch, SECRET_MASK);
>       } catch (JWTDecodeException ignored) {
> ```

add comment here describing why this exception is ignored.

File: 980-commons/src/main/java/io/harness/logging/LogSanitizerHelper.java [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+    String finalMessage = message;
+    List<String> regexMatches = pattern.matcher(message)
+                                    .results()
+                                    .map(matchResult -> finalMessage.substring(matchResult.start(), matchResult.end()))
+                                    .collect(Collectors.toList());
+    for (String regexMatch : regexMatches) {
+      try {
+        JWT.decode(regexMatch);
+        message = message.replaceAll(regexMatch, SECRET_MASK);
+      } catch (JWTDecodeException ignored) {
```
[the remark this answers is not in the document]
Comment (tiwaritejas) [reply 980-commons/src/main/java/io/harness/logging/LogSanitizerHelper.java:26]: Here, we are checking if the string that matches our regex is a valid JWT or not. If yes, we mask it. If not, it throws this exception, but that is not an actual failure, we just want to ignore the masking in that case.

**Comment by tiwaritejas** [comment]
trigger messagecheck

**Review by meenakshi-harness — APPROVED** [review]
> Reviewed 2 of 4 files at r1, 3 of 4 files at r2, 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/40549)* status: 4. unresolved discussions.

**Comment by tiwaritejas** [comment]
trigger messagecheck

**Review by mohitg0795 — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/40549)* status: :shipit: complete! ✓

**Pull request merged**